text/0000-temp-components-lib.md

```diff
@@ -0,0 +1,609 @@
+
+1. More adoption for Pando due to having a library that uses a familar pattern that people are generally more comfortable with.
+2. Should naturally teach users about the semantic token naming convention as it will inheritently use Headless-styles for prop definitions (which uses the same naming convention of the themes for
+consistency).
+3. Provide a developer experience that uses "React best practices" for a result that naturally allows scalability with custom designs via composable components, and high performance via advanced React patterns.
+4. Share the functionality as Lego blocks (create your own outcome based on what you combine/remove).
+5. By nature, it should increase the community contribution for the same reasons
+mentioned above (a familiar design).
+
+## Detailed design
+
+The design of the components library should first and foremost center around the
+Pando standards:
+
+1. A11y above all
+2. High Perfomance
+3. Easy to use
+
+Outside of these standards, there are 2 groups of components to consider:
+
+1. [Simple components](#simple-components) (Tag, Button, TextLink, etc.)
+2. [Complex compents](#complex-components) (Alert Dialogs, Notifications, Modal, etc.)
+
+Last, there should also be helpers that naturally compliment the experience:
+
+1. [Custom hooks](#custom-hooks)
+2. [Helper functions](#helper-functions)
+
+_:bulb: In addition to React 18, all components should be natively compatible for use
+within React Server Components._
+
+### Simple Components
+
+Simple components should mimick the vanilla JSX design React provides with the Headless-styles
+props extending it (i.e. not manipluataing React props or creating "component
+does all" solutions).
+
+For example, a Button should be built something like this:
+
+```typescript
+import { getButtonProps } from '@pluralsight/headless-styles'
+import type { ButtonOptions } from '@pluralsight/headless-styles/types'
+import { type HTMLAttributes, type PropsWithChildren } from 'react'
+
+interface ButtonProps extends ButtonOptions, HTMLAttributes<HTMLButtonElement> {}
+
+export function Button(props: PropsWithChildren<ButtonProps>) {
+  const {
+    className,
+    disabled,
+    icon,
+    sentiment,
+    size,
+    usage,
+    ...nativeBtnProps
+  } = props
+  const pandoProps = getButtonProps({
+    className,
+    disabled,
+    sentiment,
+    size,
+    usage,
+    icon
+  })
+
+  return (
+    <button {...pandoProps.button} {...nativeProps} />
+  )
+}
+```
+
+By nature, this example uses the native React JSX design while extending the
+Headless-styles props on top of it so it can be used in a way that compliments
+and teaches users about React & Javascript due to not manipluating any props/React APIs.
+
+We are purposefully not using `forwardedRef` to prevent complexity and possible
+performance implications. This idea might need to change, but we should wait until
+we have enough evidence the example below does not work in most use cases.
```

Comment:
I would disagree with this decision to not forward refs. At least without pointing to some data on performance impacts of forwarding refs, I've been bitten many times by library components not properly forwarding refs and would recommend as a general rule if a component is exported in a library, it should accept a ref.

One big difference between `memo` and `forwardRef` is that you can `memo` as an end-user without many issues, but `forwardRef` has to be implemented internally, and from my experience building components, the cost of not implementing it has usually been worse than including it and not needing it: often external libraries (such as `react-window` or `popper.js`) require refs to work properly, and although sometimes wrapping them in a `span` is fine, sometimes it really messes up styles, for example.

Reply:
Going to go ahead and update this portion in the RFC.